### DISCUSSION TITLE
Wrong EOL in diff viewer

### DISCUSSION THREAD

**Discussion by mstv** [gitextensions/gitextensions#13314] [Q&A] [answered]
### Environment

- GitExtensions version: vNext with #11830
- Git 2.44.0.windows.1
- Microsoft Windows NT 10.0.19045.0
- .NET 8.0.7
- DPI 96dpi (no scaling)
- Portable: False

### Issue description

1. Diff viewer shows EOL as LF although files have CRLF
2. For "\ No newline at end of file", the coloring of the previous line should be re-used

![grafik](https://github.com/user-attachments/assets/1b1e835e-5993-44c5-8cb0-af5a3f75d0f2)
![grafik](https://github.com/user-attachments/assets/3e4f7de7-96a4-4b5d-b86e-6c1794c2892b)

removed from EOF again:

![grafik](https://github.com/user-attachments/assets/6208a2ba-528f-47e8-ad8c-b38e7b119caa)

### Steps to reproduce

append a (missing) newline to a (cmake) file

### Did this work in previous version of GitExtensions?

Yes, even with the current version but only for some files, as in https://github.com/gitextensions/ICSharpCode.TextEditor/commit/cc84160cb1c03f56e0434ace876dec524565dc0c or for #10697.
Similar behavior can be reproduced with GE 2.51.01 with git 2.16.2, too.

### Diagnostics

```
$ git config --show-origin core.autocrlf
file:C:/Program Files/Git/etc/gitconfig true
```

git output is not much clearer:

![grafik](https://github.com/user-attachments/assets/fe39d01b-6f77-43e1-a156-8321a0740cfc)

**Comment by gerhardol** [comment]
GE should just display the Git output after processing color infomation.
What is the contents in the Git command log, cache tab.

The Git output could of course be customized. I believe I did not find a way to not emit "\ No newline at end of file.".
Edit: it is hardcoded: https://github.com/git/git/blob/master/diff.c#L1356 (including the \n on the previous line).

**Comment by mstv** [comment]
> What is the contents in the Git command log, cache tab.

```
[2;7;35m-#%w        The day of the current week. 0 is Sunday. (0-6)[m
\ No newline at end of file[m
[2;7;34m+[m[2;7;34m#%w        The day of the current week. 0 is Sunday. (0-6)[m
```
has CRLF when copied from cache tab and pasted to BC:

![grafik](https://github.com/user-attachments/assets/8957d811-e350-4583-9f79-6b228b97d522)

**Comment by gerhardol** [comment]
I do not see any processing issue in GE. 
git diff adds "\n\ No newline at end of file.\n" without any color information.

You can view the raw output from git bash: `git diff --color=always HEAD~1 HEAD | od -h --endian=little | tail`

So is this a request to patch the git-diff output or have I missed some incorrect processing?
Otherwise this should be directed to Git. 
(This has been discussed at GE and I assume that this is known at Git too).

Edit: GE processing related to this has not changed for months.

**Comment by RussKie** [comment]
Has #11848 addressed the issue?

**Comment by mstv** [comment]
> Has #11848 addressed the issue?

unrelated
#11848 just helps in diagnosing the issue.

**Comment by gerhardol** [comment]
> unrelated #11848 just helps in diagnosing the issue.

I interpret this issue as a wish to modify the git-diff output, postprocessing the output. (I have not seen that the GE processing is incorrect.)

**Comment by gerhardol** [comment]
ee432d3c1155ca355da6253b7a94d17d455b37f8 (proposal in private branch) marks the line as a header which is not always done now
This printout is currently the only Git-diff printout with \ in the first column, it could be cleaned.
https://github.com/git/git/blob/master/diff.c#L1347
It always start with a \n and ends with a \n

**Comment by mstv** [comment]
Regarding "\ No newline at end of file":
It does not only appear at the end of the patch: 1e1b946b65669c232db8eef55efb378d41a44b82

![grafik](https://github.com/user-attachments/assets/41fa8e16-de19-4d67-8847-bd7af2d6c9a2)

**Comment by gerhardol** [comment]
> Regarding "\ No newline at end of file": It does not only appear at the end of the patch: 

Yes, Git can emit this for both old and new file.
Not easy to follow in git-diff code:
https://github.com/git/git/blob/master/diff.c#L1356

**Comment by SAPHemlock** [comment, marked as answer]
Splitting this into the two things the thread covers.

**1. "EOL shown as LF although the files have CRLF" - this is git's output, not Git Extensions' rendering.**

Git Extensions hands git's diff bytes to the editor unmodified (`FileViewerInternal.cs:291` `TextEditor.Text = text;`; the only transformation beforehand is ANSI-escape stripping), and the EOL glyph is derived by the editor from the characters actually present (`ICSharpCode.TextEditor` `LineManager.cs:322-337`, rendered in `TextView.cs:1121-1136`). The patch parser preserves a trailing `\r` on content lines (`PatchProcessor.cs:261-279`).

So what is shown depends on what git emits. Two throwaway repos on git 2.x, `git diff | od -c`:

- `core.autocrlf=true`: the blob is LF-normalised by the clean filter, so the diff contains **no CR at all** - showing `LF` is correct.
- `.gitattributes` with `* -text`: the blob keeps CRLF and the diff does contain `\r\n`, and Git Extensions then renders CRLF.

That difference is entirely git-side (index normalisation) and matches both the `type: git issue` label and @gerhardol's "I do not see any processing issue in GE", as well as the reporter's own observation that it only happens for some files. Note also that CRLF appears after pasting into another tool because `Copy` deliberately converts to worktree EOLs (`AutoCrlfExtensions.AdjustLineEndings`, used only on the clipboard paths) - which explains the seemingly contradictory result in this thread.

A viewer-side EOL rewrite would also be actively harmful: the viewer text is fed verbatim into patch generation for stage/unstage/reset-selected-lines (`FileViewer.cs:1719`, `:1782`, `:1856`), so injecting CRs would produce patches `git apply` rejects against the LF-normalised blob.

**2. `\ No newline at end of file` presentation - this looks done.**

`7b1cd3f89` "fix: Always present git inserted as Header (#11922)" now classifies any git-inserted `\` line as a header regardless of position (`DiffLineNumAnalyzer.cs:140-152`), which also stops it consuming a left/right line number. That is the `ee432d3c` proposal from this thread, and it covers the mid-diff case raised in comment 8. Asserted by `DiffLineNumAnalyzerTests.CanGetRemovedMovedDiffInfo`.

The literal "re-use the previous line's colour" is not implemented, deliberately - header classification was chosen instead. Colour also cannot be recovered from git's output, since git resets the colour before the newline and emits the marker line uncoloured:

```
ESC[31m-line3 ESC[m \n \ No newline at end of file ESC[m \n
```

So unless the colouring half is still wanted, part 2 can be closed and part 1 is upstream git behaviour. Should this be closed?

**Reply by mstv** [reply]
Thank you for the analysis!
